### DISCUSSION TITLE
Can Raven achieve the same performance as ElasticSearch?

### DISCUSSION THREAD

**Discussion by bmills74** [ravendb/ravendb#15364] [Q&A] [answered]
We've just completed a large project which analyses user's data and presents it in a filterable analytics pie charts.  The way I structured the indexes was using auto indexes that grouped all the data in many efficient auto indexes using documentquery eg:
 
Gender chart:
```
 var genderQuery= analyticsSession.Advanced.DocumentQuery<T>()
                    .GroupBy("Gender", "Type",
                       "Date"  )        
                       .SelectKey(chartKey, "ChartKey")
                       .SelectKey("Type", "Type")
                       .SelectKey("Date", "Key1")
                       .SelectKey("Gender", "Key2")
                      .SelectCount()
                      .OfType<ReduceResult>()
```
The above allows for efficient grouping and an index without too many entries and it's very quick to load.  The problem is that our clients have around 20 fields that they want to display as well as filter any combination.  So the autoindex that's used for this has 20 groupings which means there is no efficient grouping, with the number of index entries almost the same as the actual number of documents.  This means that when a worst case scenario happens and they filter on the biggest collection with a wide net, it's retrieving 500k plus index entries, taking over a minute to retrieve and using 500mb of bandwidth in one query! Very gross.

We decided to give elasticsearch a try and I've been astonished at how quick it is to run the same set of filters on the same size collection, just a second or two on a worst case scenario. Looking at the index it seems it's able to achieve this using just one small index.
Rewriting the project to use ElasticSearch is a large undertaking and before I embark on this, I'm wanting to check if there's something I've missed with Raven that could achieve the same kind of performance as Elastic, or is there a fundamental difference in the db structure that makes it impossible?

**Comment by ayende** [comment, marked as answer]
Hi,

`GroupBy` is to generate a map/reduce index. And if you are grouping by a lot of individual values, which constantly change, a better way to manage that is to use facets. See: https://ravendb.net/docs/article-page/5.4/csharp/indexes/querying/faceted-search#step-3

The `AggregateBy` gives RavenDB more freedom in how it works and should match more closely what you are doing in Elastic.

If there is still an issue, please post a full scenario that we can look at.

**Comment by bmills74** [comment]
Initial testing using facets have gone very well and are also incredibly quick! Wish I'd known about them before, but better late than never.  Thanks very much for the help @ayende . :-)
